src/mlir/consumer.cc

```diff
@@ -0,0 +1,762 @@
+      if (classField)
+      {
+        auto [subPos, subTy, subFound] = getField(classField, fieldName);
+        pos += subPos;
+
+        if (subFound)
+          return {pos, subTy, true};
+        else
+          continue;
+      }
+
+      // Not found, continue searching
+      pos++;
+    }
+
+    // Not found, return false
+    return {0, Type(), false};
+  }
+
+  // ===================================================== AST -> MLIR
+  llvm::Error ASTConsumer::consumeRootModule(Ast ast)
+  {
+    auto node = nodeAs<Class>(ast);
+    assert(node && "Bad node");
+
+    // TODO: Quick early pass to declare all functions before we descend the
+    // AST.
+
+    // Modules are just global classes
+    return consumeClass(ast);
+  }
+
+  llvm::Error ASTConsumer::consumeClass(Ast ast)
+  {
+    auto node = nodeAs<Class>(ast);
+    assert(node && "Bad node");
+
+    StringRef modName = node->id.view();
+    // The root module has no name, doesn't need to be in the context
+    if (!modName.empty())
+      functionScope.push_back(modName);
+
+    // Push another scope for variables, functions and types
+    SymbolScopeT var_scope(symbolTable());
+    auto type = StructType::getIdentified(builder().getContext(), modName);
+
+    // Create an entry for the fields
+    classFields.emplace(modName, FieldOffset());
+
+    // Lower members, types, functions
+    for (auto sub : node->members)
+    {
+      switch (sub->kind())
+      {
+        case Kind::Class:
+        {
+          auto err = consumeClass(sub);
+          if (err)
+            return err;
+          functionScope.pop_back();
+          break;
+        }
+        case Kind::Using:
+          // Ignore for now as this is just a reference to the module name
+          // that will be lowered, but module names aren't being lowered
+          // now.
+          break;
+        case Kind::Function:
+        {
+          auto func = consumeFunction(sub);
+          if (auto err = func.takeError())
+            return err;
+          gen.push_back(*func);
+          break;
+        }
+        case Kind::Field:
+        {
+          auto res = consumeField(sub);
+          if (auto err = res.takeError())
+            return err;
+
+          // Update class-field map
+          auto field = nodeAs<Field>(sub);
+          auto& fieldMap = classFields[modName];
+          fieldMap.fields.push_back(field->location.view());
+          fieldMap.types.push_back(*res);
+          break;
+        }
+        default:
+          return runtimeError("Wrong member in class");
+      }
+    }
+    if (mlir::failed(
+          type.setBody(classFields[modName].types, /*packed*/ false)))
+      return runtimeError("Error setting fields to class");
+
+    return llvm::Error::success();
+  }
+
+  llvm::Expected<Value> ASTConsumer::consumeNode(Ast ast)
+  {
+    switch (ast->kind())
+    {
+      case Kind::Lambda:
+        return consumeLambda(ast);
```

Comment:
This is a pattern that came from the old AST and I didn't optimise because I wasn't sure how long we'd have the AST like that. It seems long, now, so I'll do that work soon.